### PR TITLE
destination: Fix GetProfile endpoint buffering

### PR DESCRIPTION
**Pull request by olix0r**
In 71635cb and 357a1d3 we updated the endpoint and profile translators to prevent backpressure from stalling the informer tasks. This change updates the endpoint profile translator with the same fix, so that updates are buffered and can detect when when a gRPC stream is stalled.

Furthermore, the update method is updated to use a protobuf-aware comparison method instead of using serialization and string comparison.

A test is added for the endpoint profile translator, since none existed previously.

### CODE REVIEW

File: controller/api/destination/endpoint_profile_translator.go [hunk 1]
```diff
@@ -1,69 +1,136 @@
...
+	}
+	return true, nil
+}
+
+func (ept *endpointProfileTranslator) update(address *watcher.Address) {
 	opaquePorts := watcher.GetAnnotatedOpaquePorts(address.Pod, ept.defaultOpaquePorts)
 	endpoint, err := ept.createEndpoint(*address, opaquePorts)
 	if err != nil {
-		return false, fmt.Errorf("failed to create endpoint: %w", err)
+		ept.log.Error(err)
```
Comment (alpeb) [inline controller/api/destination/endpoint_profile_translator.go:114]: Before this change no message was sent back when an error occurred; now we're sending a profile with an empty Endpoint. Is that intended?

File: controller/api/destination/endpoint_profile_translator.go [hunk 2]
```diff
@@ -1,69 +1,136 @@
...
+// endpointProfileUpdatesQueueOverflowCounter is a prometheus counter that is incremented
+// whenever the profile updates queue overflows.
+//
+// We omit ip and port labels because they are high cardinality.
+var endpointProfileUpdatesQueueOverflowCounter = promauto.NewCounter(
+	prometheus.CounterOpts{
+		Name: "endpoint_profile_updates_queue_overflow",
+		Help: "A counter incremented whenever the endpoint profile updates queue overflows",
+	},
+)
```
Comment (alpeb) [inline controller/api/destination/endpoint_profile_translator.go:38]: No label for service?

File: controller/api/destination/endpoint_profile_translator.go [hunk 3]
```diff
@@ -1,69 +1,136 @@
...
+		select {
+		case <-ept.endStream:
+			// The endStream channel has already been closed so no action is
+			// necessary.
+		default:
+			ept.log.Error("Profile update queue full; aborting stream")
+			close(ept.endStream)
+		}
+	}
+	return true, nil
```
Comment (alpeb) [inline controller/api/destination/endpoint_profile_translator.go:107]: The pod watcher calls this function whenever the pod changes. The boolean returned is supposed to tell whether that change implied sending an update back into the stream, in which case the pod watcher will increment the `ip_port_updates` metric. Always returning true here will inflate that accounting.

File: controller/api/destination/test_util.go [hunk 4]
```diff
@@ -464,7 +464,7 @@ spec:
 		t.Fatalf("NewFakeMetadataAPI returned an error: %s", err)
 	}
 	log := logging.WithField("test", t.Name())
-	logging.SetLevel(logging.TraceLevel)
+	// logging.SetLevel(logging.TraceLevel)
```
Comment (alpeb) [inline controller/api/destination/test_util.go:467]: Ah I think I might have left this here by mistake.

File: controller/api/destination/endpoint_profile_translator_test.go
Comment (alpeb) [inline controller/api/destination/endpoint_profile_translator_test.go:1]: Very nice :+1:

File: controller/api/destination/endpoint_profile_translator.go (hunk 2 above)
Comment (olix0r) [reply controller/api/destination/endpoint_profile_translator.go:38]: These lookups are specifically for non-service IPs.

Comment (alpeb) [reply controller/api/destination/endpoint_profile_translator.go:40]: It's also used when getting profile by name, so yeah that label won't make sense in all cases.

File: controller/api/destination/endpoint_profile_translator.go (hunk 3 above)
Comment (olix0r) [reply controller/api/destination/endpoint_profile_translator.go]: I've updated the Update interface so now that we only return an error -- with the buffer we can't actually know whether the consumer sends the update or not. We now return an error when the address can't be enqueued.

**Review by alpeb — APPROVED** [review]

**Pull request merged**